### PR TITLE
Update Sidebar - Expanding Button Section

### PR DESCRIPTION
**Pull request by tlesicka**
Continuation of splitting https://github.com/actualbudget/actual/pull/3457 into several PR's as requested by @MatissJanis.

This PR updates the expanding "More" section to keep all buttons in the same flow instead of being divided by the "More".  
"Settings" is removed from the expanding buttons section as it can be accessed from the budget name drop down menu and does not need to be in two places.

Previous discussions can be found:
Issue #3414
PR #3457

### CODE REVIEW

File: packages/desktop-client/src/components/sidebar/PrimaryButtons.tsx [hunk 1]
```diff
@@ -1,72 +1,96 @@
...
-  SvgCheveronRight,
-  SvgCog,
-  SvgReports,
-  SvgStoreFront,
-  SvgTuning,
-  SvgWallet,
-} from '../../icons/v1';
-import { SvgCalendar } from '../../icons/v2';
+import { useLocalPref } from '../../hooks/useLocalPref';
+import { SvgCheveronDown, SvgCheveronUp } from '../../icons/v1';
```
[the remark this answers is not in the document]
Comment (tlesicka) [reply packages/desktop-client/src/components/sidebar/PrimaryButtons.tsx:11]: This is the way the icons are spelled in icons/v1.  Other components use the icons with this spelling, so I will not change it in this PR.

File: packages/desktop-client/src/components/sidebar/PrimaryButtons.tsx [hunk 2]
```diff
@@ -2,71 +1,96 @@
...
+  };
+
   useEffect(() => {
-    if (isActive) {
-      setOpen(true);
+    if (divRef?.current && itemsRef?.current) {
+      divRef.current.style.height = itemsRef.current.scrollHeight + 'px';
     }
-  }, [isActive, location.pathname]);
+  }, [expanded, location.pathname]);
```
[the remark this answers is not in the document]
Comment (tlesicka) [reply packages/desktop-client/src/components/sidebar/PrimaryButtons.tsx:52]: divRef is set the the height of the inner container itemsRef.  itemsRef will change size based off how many items are inside.  This makes the transition work correctly.  If I set the height to 0px then no buttons would show.

File: packages/desktop-client/src/components/sidebar/PrimaryButtons.tsx [hunk 3]
```diff
@@ -2,71 +1,96 @@
...
 
 import { Item } from './Item';
 import { SecondaryItem } from './SecondaryItem';
 
-export function PrimaryButtons() {
+type PrimaryButtonItems = {
+  title: string;
+  Icon:
+    | ComponentType<SVGProps<SVGElement>>
+    | ComponentType<SVGProps<SVGSVGElement>>;
```
[the remark this answers is not in the document]
Comment (tlesicka) [reply packages/desktop-client/src/components/sidebar/PrimaryButtons.tsx:22]: Updated as indicated

**Comment by Jonathan-Fang** [comment]
@jfdoming 's https://github.com/actualbudget/actual/pull/3648 gave me an idea for organizing the sidebar.

Although it's currently set to the top right, I proposed two ideas.

1. Place the help button above the add account button, having tiny buttons for settings and help in the sidebar below the accounts
unified help experience
Please excuse my MS Paint 😅 

So in order it would be
- Accounts
- then smaller font, settings, help
- then the add account button.

3. Either placing the help button under the More dropdown, along with Settings in the smaller font.
    In my opinion, settings can be grouped separately away from rules and payees.
    If rules and payees remain in the More accordion, then Settings and Help can be above the add account button.

Thoughts?

**Comment by MatissJanis** [comment]
👋 I don't quite understand **why** we would need to change this part of the UI? Personally the new proposal here seems a lot less polished than what we currently have in `edge`.

**Comment by tlesicka** [comment]
Personally, I don't think there should be a "More" section on the sidebar.  In both this PR and ```edge``` the "More" section disrupts the simple flow of the app.  IMO, the items in the More section should be moved to the menu and the More section should be removed.
However, other people (as discussed in #3414) like having Payees and Rules in the More section.  I see this as a potential future problem.  As more things get added to AB, this More section is going to expand until it is too cluttered.
What should be placed in the More section vs what should be placed in the menu?
On ```edge```, Settings is duplicated in both the menu and the More section.  This should only be in one place.  My thought is that it should be removed from the More section and left in the menu, as in this PR.  Or placed as just a settings cog icon with a hover text, like the new help menu.
The other part that I'm working on (multi-currency) will require a page for Exchange Rates.  I believe this as well as Payees and Rules should go under the menu, not under More.  Is Exchange Rates less important than Rules or Payees?  IMO, no, they are all the same.  Are these three less important than Budget and Reports?  Yes, so I agree that Budgets and Reports should always be visible.  So, with the current layout, Exchange Rates would be added to the sidebar, which will add clutter.  This is why I would prefer to remove the More section completely. (But, ultimately, where we put the Exchange Rates page is a discussion for another time.  I'm just using it as an example.)

IMO the current More section disrupts the look and flow of the sidebar.  While I agree that Payees, Rules, and Settings are not as important as Budget and Reports, I don't believe that they should be a smaller button size.  That is why this PR uses the sliding/expanding More section to reveal more items.  I am not suggesting this next part, but mentioning it as an example.  If Reports was to be an expanding section, then yes, whatever is underneath Reports should be a smaller button style.  But, Payees, Rules, and Settings are not underneath any other section, so IMO they should not be smaller than the rest.

I don't understand what you mean by this PR is less polished than ```edge```.

On ```edge``` during a page refresh, the expanding More section acts different based on what is active.  If Budget, Reports, or Schedules is active and More is expanded, during a page refresh the More section collapses.  If Rules, Payees, or Settings is active and More is collapsed, then More expands on the page refresh.
In this PR, the state stays the same across page refresh.  If More is expanded it stays expanded during refresh and if it is collapsed it stays collapsed.
On ```edge``` when Payees, Rules, or Settings is active and the More section is closed then More gets highlighted to show that something is active inside the More section.  In this PR, that page button stays active and stays on the top level.

From a code level, currently, the Primary Buttons/expanding section are hardcoded into the PrimaryButtons component.  For cleaner code and better maintainability, it would be better to have the PrimaryButtons component read a list of buttons that is given to it from the main Sidebar component.  I know that the previous PR that I presented and that was merged separated the PrimaryButtons into it's own component and that was intentional.  Yes, PrimaryButtons should be it's own component, but the actual content should be managed from the main Sidebar.  This is the way that SecondaryButtons was merged.  The main menu content should also be separated from the BudgetName component even though it is displayed on the BudgetName component (but, that's for a different PR).  This allows easy maintainability by having menu items, primary buttons, and secondary buttons created as lists in the Sidebar component, while the actual code to display those items is generated from the individual components (ie modular design).

I understand that there are many people who have worked on this project in the past and all have strong opinions, which is not necessarily a bad thing.  I highly appreciate and respect all of you for bringing AB to the place that it is.  However, it still has a bit of a rugged and unfinished look to it.  I think there are quite a few people moving from other software that are used to a more finished look and feel.  I know that I am one of those people and while I don't want to clone those other apps, I think that seeing what works for other software is a good way to make this app better.  Neither YNAB4 nor nYNAB use an expanding more section, and both have a better menu system.
So I would suggest changing this PR to remove the More section and incorporate those items into the menu.

**Comment by Jonathan-Fang** [comment]
I agree with @tlesicka 's notes on wanting to get AB a more polished UI, and I think the "ruggedness" is more like the current misalignment of the More toggle and everything below it, it doesn't line up with the rest of the items in the sidebar.

A big question here is where to put Payees, Rules, and a pending Exchange Rates button. At this point I may even suggest adding them at the bottom of the sidebar in a slightly smaller font because they're only three buttons. And the top part of the sidebar can be the more prominent Budget, Reports, and Schedules, with a menu to rename the budget or even prepare for multi-user.

Another idea is to put the Payees, Rules, and the pending Exchange Rates button inside the menu dropdown when you click the Budget Name, and then everything is in one place. The proposed PR to create the react command bar would make accessibility quite easy, but even for click-only users, knowing that there's a smooth workflow to navigating the software is key to make Actual Budget appeal to everyone, not just developers.

**Comment by tlesicka** [comment]
While this is well beyond the scope of this PR, here's an idea for a way to clean up the side bar and prepare for possible future additions.

top bar idea

Not all elements would be active when the topbar is new, but the idea is to have a plan in place to where future things will go.  The center input box where Cmd-K will eventually live would not be visible until that feature is built.  The budget name there was just to add initial text.  I don't care if it's the budget name or "search" or whatever wording, that will be for whoever builds that component.

This would remove the menu dropdown from the budget name.  (Some people have mentioned that it is odd to have the budget name as a dropdown menu.)
The File menu would have import/export, backup, close file, settings, etc.  Possibly place Payees, Rules, Exchange Rates there too.
The Edit menu would have your standard undo/redo or possibly place Payees, Rules, Exchange Rates here.
Actual Logo would be a link to actualbudget.org

Again, this is just a quick thrown together idea, highly based off of VS Code and will need a lot of tweaks to get the visuals to look great.  But, I think it would be good to have a plan of where the design is going instead of just putting the pieces together as someone develops a piece.

**Comment by MatissJanis** [comment]
👋 I really appreciate the work you're doing on improving the nav, but sadly here I will have many conflicting opinions. 

Before we begin - here are two good reads for how we think about design in Actual

- https://actualbudget.org/docs/contributing/#the-design-strategy-of-actual
- https://actualbudget.org/docs/vision

Let me go through things one by one.

> However, other people (as discussed in https://github.com/actualbudget/actual/issues/3414) like having Payees and Rules in the More section. I see this as a potential future problem. As more things get added to AB, this More section is going to expand until it is too cluttered.

Generally we try to follow the [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) principle here. Currently we don't have the bloat in side-nav items, so the problem does not exist. If we ever get to a point where there are too many items - we can return to solving the problem. But until then - lets not over-complicate things.

> Settings is duplicated in both the menu and the More section.

I am open to the menu settings item being removed.

> The other part that I'm working on (multi-currency) will require a page for Exchange Rates. I believe this as well as Payees and Rules should go under the menu, not under More.

Exchange rates is not a first-party feature of Actual. Similar to bank-sync: users can discover the feature when they need to use it. As such: it could be a modal or separate page, but it should not use primary real-estate in the side nav.

> I don't understand what you mean by this PR is less polished than edge.

The margins/paddings and font size are inconsistent and not aligned.

> In this PR, the state stays the same across page refresh.

This is a good functional change. I have no objections here.

> From a code level [..]

Honestly: I've not reviewed the code. My focus here so far has been on the product and design implications. Code review comes next. On a high-level I think I won't have any objections about the implementation (but again: I've not actually reviewed it).

> However, it still has a bit of a rugged and unfinished look to it. [..]

I would argue that inconsistent margins/paddings and mis-aligned elements makes the app look even more unfinished and rugged. But taking a step back: I would still like to understand the "why" behind this change. What does the product gain from this change?

---

> While this is well beyond the scope of this PR, here's an idea for a way to clean up the side bar and prepare for possible future additions.

Interesting proposal, but lets discuss it first separately before building it.

**Comment by Jonathan-Fang** [comment]
@MatissJanis Although your question was directed to @tlesicka, I would also like to contribute to the discussion on the reason for this PR change.

IMO, the current top-section of the sidebar as being discussed in this PR is not "sleek and clutter-free" as stated in https://actualbudget.org/docs/vision/.

image

First, there are two competing menus [the More toggle and the dropdown budget menu] that have secondary priority items (Payees, Rules, Settings, and Help), and I recommend choosing one place if the app is supposed to be clutter-free.
- You have mentioned you agree with removing the redundancy of the Settings button, so the question now is to double down on where it should live.

Second, the Payees and Rules feel like an afterthought inside the More cabinet; it's the items inside that look misaligned compared to everything else in the sidebar.

I agree with your statement that "inconsistent margins/paddings and mis-aligned elements makes the app look even more unfinished" because that's what the this top section of the sidebar in its current state (not this PR version) looks like to me. The PR version actually looks better because it attempts to address the connection between the More toggle and the items underneath.

That being said, if it's this accordion setup that is disagreeable, I propose keeping the way @tlesicka has the Payees and Rules be the same button font, margins and padding as the Budget and Reports, but remaining in the More toggle.

It's hide-able behind a toggle anyway, and the indented subitems is enough for me to indicate its relationship as a subitem to the mother toggle.

I welcome any feedback on this take and comments.

**Comment by tlesicka** [comment]
@MatissJanis I looked back at the changes that have happened to the sidebar and specifically the expanding More section.  Now I understand your perspective a little better.  Or at least, I think I understand it a little better.  There has already been some cleanup of the sidebar and there was a move away from the menu in favor of putting more items on the sidebar.  While that is a good thought in finding things that would otherwise be hidden in a menu, it also tends to clutter things up.  You, having seen things before and now the way they are, are looking at this from an improvement standpoint.  And I get that.  But, it still doesn't have the final polish that in my opinion it could and should have.  In my opinion it still looks a little hacked together.

So, to answer your primary question:
> I would still like to understand the "why" behind this change. What does the product gain from this change?

This change was initially proposed to remove the More section completely.  I agree with you when you said:
> I would argue that inconsistent margins/paddings and mis-aligned elements makes the app look even more unfinished and rugged.

This is the "why" behind this PR.  The current more section in both the expanded and collapsed state is not sleek nor clutter-free and looks mis-aligned, even though it is more clutter-free than what it was.
- Secondary icons are located in the blank space of the primary icon and primary name.
- Spacing between More and secondary items more closely matches the spacing between primary items instead of the spacing between secondary items.
- When selecting the settings from the menu, the more section opens.
- Why do we have both a drop down menu and an expanding more section?  Even from looking at the previous PR's, I'm still not sure.

That is why in #3414 I initially proposed removing it and assigning the Payees and Rules as menu items instead of action buttons.  

I think I understand now that removing the expanding More section is not going to happen, even though I (and I believe others too) would think that this would clean up the sidebar.  If I am mistaken in my understanding of your position, let me know.  I truly want to make this product look amazing.


> The margins/paddings and font size are inconsistent and not aligned.

Thank you for letting me know this.  I looked at it and you are correct.  I adjusted it a bit at the button level.  The space between the icon and the text needs to be adjusted but I opted to not work on that until we have a final consensus on the status of this PR.  Also, the font size can be brought up to a full size button, that is not a big deal.

**Comment by MatissJanis** [comment]
Apologies for my bluntness:

I still do not see a good argument for moving the "more" link around. From a design perspective - I still think our current implementation has an edge over the proposal here (which still has paddings/margins all over the place, mismatching fonts, etc.). On a functional level - the "settings" link **must** remain in the side-nav.

Because of these three things: I will not be approving this PR.

**Pull request closed without merging**